File: fedorova-t.v-kt-41-22/Database/Configurations/DepartmentConfiguration.vb
Imports fedorova_t.v_kt_41_22.Models
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Metadata.Builders


Namespace fedorova_t.v_kt_41_22.Database.Configurations
    Public Class DepartmentConfiguration
        Implements IEntityTypeConfiguration(Of Department)
        Public Sub Configure(builder As EntityTypeBuilder(Of Department)) Implements IEntityTypeConfiguration(Of Department).Configure
            builder.ToTable("Departments")
            builder.HasKey(Function(d) d.Id)
            builder.Property(Function(d) d.Name).HasMaxLength(100).IsRequired()

            builder.HasOne(Of Teacher)(Function(d) CType(d.Head, Teacher?)).WithOne(Function(t) CType(t.ManagedDepartment, Department?)).HasForeignKey(Of Department)(CType(Function(d) CObj(d.HeadId), Expressions.Expression(Of Func(Of Department, Object?)))).OnDelete(DeleteBehavior.Restrict)
        End Sub
    End Class
End Namespace

File: fedorova-t.v-kt-41-22/Database/Configurations/DepartmentConfiguration.cs
using fedorova_t.v_kt_41_22.Models;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
using fedorova_t.v_kt_41_22.Database.Helpers;


namespace fedorova_t.v_kt_41_22.Database.Configurations
{
    public class DepartmentConfiguration : IEntityTypeConfiguration<Department>
    {
        private const string TableName = "Departments";
        public void Configure(EntityTypeBuilder<Department> builder)
        {
            builder.HasKey(p => p.Id)
            .HasName($"pl_{TableName}_department_id");

            //Автоинкрементация
            builder.Property(p => p.Id)
                .ValueGeneratedOnAdd()
                .HasColumnName("Department_Id")
                .HasComment("Id факультета");

            builder.Property(p => p.Name)
                 .IsRequired()
                 .HasColumnName("Name")
                 .HasColumnType(ColumnType.String).HasMaxLength(20)
                 .HasComment("Название факультета");

            builder.Property(p => p.FoundedDate)
                 .IsRequired()
                 .HasColumnName("Founded_Date")
                 .HasColumnType(ColumnType.Date)
                 .HasComment("Дата основания факультета");

            // Настройка связи
            builder.Property(p => p.HeadId)
              .HasColumnName("Head_Id")
              .HasComment("Id зав. кафедры");


            builder.HasOne(p => p.Head)
                .WithOne()
                .HasForeignKey<Department>(p => p.HeadId)
                .OnDelete(DeleteBehavior.NoAction)
                .HasConstraintName("fk_f_head_id");


            builder.ToTable(TableName)
                .HasIndex(p => p.HeadId, $"idx_{TableName}_fk_f_head_id");
        }
    }
    }
